adr/ADR-47.md

```diff
@@ -0,0 +1,115 @@
+
+## Config
+
+### Total timeout
+
+The maximum amount of time to wait for responses. When the time is expired, the process is complete.
+The wait for the first message is always made with the total timeout since at least one message must come in within the total time.
+
+* Always used
+* Defaults to the connection or system request timeout.
+
+### Stall timer
+
+The amount time to wait for messages other than the first (subsequent waits). 
+Considered "stalled" if this timeout is reached, indicating the request is complete.
+
+* Optional
+* Less than 1 or greater than or equal to the total timeout behaves the same as if not supplied.
+* Defaults to not supplied.
+* When supplied, subsequent waits are the lesser of the stall time or the calculated remaining time. 
+This allows the total timeout to be honored and for the stall to not extend the loop past the total timeout.
+
+### Max messages
+
+The maximum number of messages to wait for. 
+* Optional
+* If this number of messages is received, the request is complete.
+* If this number is supplied and total timeout is not set, total timeout defaults to the connection or system timeout.
+
+### Sentinel
+
+While processing the messages, the user should have the ability to indicate that it no longer wants to receive any more messages.
+* Optional
+* Language specific implementation
+* If sentinel is supplied and total timeout is not set, total timeout defaults to the connection or system timeout.
+
+## Notes
+
+### Message Handling
+
+Each client must determine how to give messages to the user.
+* They could all be collected and given at once.
+* They could be put in an iterator, queue, channel, etc.
+* A callback could be made.
+
+### End of Data
+
+The developer should notify the user when the request has stopped processing and the receiving mechanism is not fixed like a list
+or iterator that termination is obvious. A queue or a callback for instance, should get a termination message.
+Implementation is language specific based on control flow.
+
+### Status Messages / Server Errors
+
+If a status (like a 503) or an error comes in place of a user message, this is terminal.
+This is probably useful information for the user and can be conveyed as part of the end of data.
+
+#### Callback timing
+
+If callbacks are made in a blocking fashion, 
+the client must account for the time it takes for the user to process the message 
+and not consider that time against the timeouts.
+
+### Sentinel
+
+If the client supports a sentinel with a callback/predicate that accepts the message and returns a boolean, 
+a return of true would mean continue to process and false would mean stop processing.
+
+If possible, the client should support the "standard sentienl", which is a message with a null/nil or empty payload.
+
+### Cancelling
+
+If possible, the user should be able to cancel the request. This is not the sentinel.
```

Comment:
here for example would mean to unsub without a limit set

Reply:
This is intended that the dev can cancel the entire request-many arbitrarily. For instance I could envision a request that gives 1 second for responses, gets many, realizes it has all the information it needs and is done. If it's using a manual sentinel pattern with a predicate, they can short circuit that way, but if they are not using a sentinel, this is just another way for manually short circuit.